test/ProductControllerTest.scala

```diff
@@ -33,4 +39,98 @@ class ProductControllerTest extends PlaySpec with Results {
+      val result: Future[Result] = controller.getInventory().apply(FakeRequest())
+      val json : JsValue = Json.parse(contentAsString(result))
+
+      val existingId = (json.as[Seq[JsObject]].head \ "id").as[String]
+      val noneId = existingId.replace(existingId.charAt(0), (existingId.charAt(0).toInt - 1).toChar)
+      val invalidId = ""
+
+      status(controller.getProduct(existingId).apply(FakeRequest())) mustBe 200
+      status(controller.getProduct(noneId).apply(FakeRequest())) mustBe 404
+      status(controller.getProduct(invalidId).apply(FakeRequest())) mustBe 404
+    }
+  }
+
+  implicit lazy val materializer: Materializer = app.materializer
+  "POST <root>/rest/bakery/product test" should {
+    "be valid" in {
+      val environment = Environment.simple()
+      val controller = new ProductController(Helpers.stubControllerComponents(), environment)
+      val requestJson = Json.obj(
+        "name" -> "Kunafa",
+        "quantity" -> 2,
+        "price" -> 12.5
+      )
+
+      val request = FakeRequest(POST, "rest/bakery/product").withHeaders(CONTENT_TYPE -> "application/json")
+        .withBody(requestJson)
+
+      val result = call(controller.addProduct(), request)
+      status(result) mustBe 200
+    }
+  }
+
+  "PUT <root>/rest/bakery/product/:id test" should {
+    "be valid" in {
+      val environment = Environment.simple()
+      val controller = new ProductController(Helpers.stubControllerComponents(), environment)
+
+      val result: Future[Result] = controller.getInventory().apply(FakeRequest())
+      val json: JsValue = Json.parse(contentAsString(result))
+      val element = json.as[Seq[JsObject]].head
+
+      val existingId = (element \ "id").as[String]
+      val noneId = existingId.replace(existingId.charAt(0), (existingId.charAt(0).toInt - 1).toChar)
+      val invalidId = ""
+
+      val requestJson = Json.obj(
+        "name" -> (element \ "name").as[String],
+        "quantity" -> ((element \ "quantity").as[Int] + 1),
+        "price" -> ((element \ "price").as[Double] + 1)
+      )
+
+      val requestGood = FakeRequest(PUT, "rest/bakery/product/"+existingId).withHeaders(CONTENT_TYPE -> "application/json")
+        .withBody(requestJson)
+
+      val requestNoId = FakeRequest(PUT, "rest/bakery/product/" + noneId).withHeaders(CONTENT_TYPE -> "application/json")
+        .withBody(requestJson)
+
+      val requestInvalid = FakeRequest(PUT, "rest/bakery/product/" + invalidId).withHeaders(CONTENT_TYPE -> "application/json")
+        .withBody(requestJson)
+
+      val resultGood = call(controller.updateProduct(existingId), requestGood)
+      val resultNoId = call(controller.updateProduct(noneId), requestNoId)
+      val resultInvalid = call(controller.updateProduct(invalidId), requestInvalid)
+
+      status(resultGood) mustBe 200
+      status(resultNoId) mustBe 404
+      status(resultInvalid) mustBe 404
+    }
+  }
```

Comment:
separate use cases. Each test should only be expecting a single http result, lumping all of those makes this test encompass more than it needs to and makes it hard to debug and visualize